modules/ipc/include/hephaestus/ipc/zenoh/service.h

```diff
@@ -4,30 +4,146 @@
 
 private:
   SessionPtr session_;
   std::unique_ptr<zenohc::Queryable> queryable_;
 
-  std::string topic_;
+  TopicConfig topic_config_;
   Callback callback_;
 };
 
+template <typename RequestT, typename ReplyT>
+auto callService(const Session& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::chrono::milliseconds& timeout) -> ReplyT;
+
+// --------- Implementation ----------
+
+template <typename RequestT, typename ReplyT>
+constexpr void checkTemplatedTypes() {
+  static_assert(serdes::ProtobufSerializable<RequestT> || std::is_same_v<RequestT, std::string>,
+                "Request needs to be serializable or std::string.");
+  static_assert(serdes::ProtobufSerializable<ReplyT> || std::is_same_v<ReplyT, std::string>,
+                "Reply needs to be serializable or std::string.");
+}
+
+template <class RequestT>
+auto createRequest(const zenohc::Query& query) -> RequestT {
+  if constexpr (std::is_same_v<RequestT, std::string>) {
+    return static_cast<std::string>(query.get_value().as_string_view());
+  } else {
+    RequestT request;
+    auto payload = query.get_value().get_payload();
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+    serdes::deserialize<RequestT>(buffer, request);
+    return request;
+  }
+}
+
+template <typename RequestT, typename ReplyT>
+Service<RequestT, ReplyT>::Service(SessionPtr session, TopicConfig topic_config, Callback&& callback)
+  : session_(std::move(session)), topic_config_(std::move(topic_config)), callback_(std::move(callback)) {
+  checkTemplatedTypes<RequestT, ReplyT>();
+
+  auto query = [this](const zenohc::Query& query) mutable {
+    DLOG(INFO) << fmt::format("Received query from '{}'", query.get_keyexpr().as_string_view());
+
+    auto reply = this->callback_(createRequest<RequestT>(query));
+    zenohc::QueryReplyOptions options;
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_APP_CUSTOM });  // Update encoding.
+      query.reply(this->topic_config_.name, reply, options);
+    } else {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
+      query.reply(this->topic_config_.name, serdes::serialize(reply), options);
+    }
+  };
+
+  queryable_ = expectAsUniquePtr(
+      session_->zenoh_session.declare_queryable(topic_config_.name, { std::move(query), []() {} }));
+}
+
+template <typename RequestT, typename ReplyT>
+auto callService(const SessionPtr& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::chrono::milliseconds& timeout) -> std::optional<ReplyT> {
+  checkTemplatedTypes<RequestT, ReplyT>();
+
+  DLOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
```

Comment:
I haven't tried myself yet, but this could be a good use case to use `absl::Mutex` which comprise of a mutex plus a condition variable internally so we do not need two separate variable and all locking overhead

Reply:
I don't think so, `absl::Mutex::Wait` still needs a condition if I understand correctly, so not sure if it would make things easier.